nixos/doc/manual/development/modular-services.md

```diff
@@ -0,0 +1,91 @@
+
+# Modular Services {#modular-services}
+
+Status: in development. This functionality is new in NixOS 25.05, and significant changes should be expected. We'd love to hear your feedback in <!-- FIXME PR link -->
+
+Traditionally, NixOS services were defined using sets of options *in* modules, not *as* modules. This made them non-modular, resulting in problems with composability, reuse, and portability.
+
+A *modular service* is a [module] that defines values for a core set of options, including which program to run.
+
+NixOS provides two options into which such modules can be plugged:
+
+- `system.services.<name>`
+- an option for user services (TBD)
+
+Crucially, these options have the type [`attrsOf`] [`submodule`].
+The name of the service is the attribute name corresponding to `attrsOf`.
+<!-- ^ This is how composition is *always* provided, instead of a difficult thing (but this is reference docs, not a changelog) -->
+The `submodule` is pre-loaded with two modules:
+- a generic module that is intended to be portable
+- a module with systemd-specific options, whose values or defaults derive from the generic module's option values.
+
+So note that the default value of `system.services.<name>` is not a complete service. It requires that the user provide a value, and this is typically done by importing a module. For example:
+
+<!-- Not using typical example syntax, because reading this is *not* optional, and should it should not be folded closed. -->
+```nix
+{
+  system.services.httpd = {
+    imports = [ nixpkgs.modules.services.foo ];
+    foo.settings = {
+      # ...
+    };
+  };
+}
+```
+
+## Portability {#modular-service-portability}
+
+It is possible to write service modules that are portable. This is done by either avoiding the `systemd` option tree, or by defining process-manager-specific definitions in an optional way:
+
+```nix
```

Comment:
This needs some verbal annotation that this is a *service module*, because given the introduction it's not easy to figure out just from the `_class`. Which is why I'd argue we should use actual examples, since they support free-form annotations.

Reply:
> free-form annotations

Haven't heard of that. Can't find anything about examples, markdown or commonmark in https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md

Reply:
https://github.com/NixOS/nixpkgs/blob/master/doc/README.md#admonitions

By annotation I mean simply freeform text contained in the example block. And again, that examples are collapsed by default is simply a misfeature and I've seen it in different places on the web recently, no clue why people would by default ignore the most important parts of documentation.